### PR TITLE
Finish step 3.11 of the Rust Incubator

### PR DESCRIPTION
**Pull request by ffuugoo**
Finished step 3.11 (Async I/O, futures and actors) of the Rust Incubator

- Read documentation listed in README of the step
- Implemented a html-file-downloader-program according to task

### CODE REVIEW

File: 3_ecosystem/3_11_async/src/main.rs [hunk 1]
```diff
@@ -41,46 +51,70 @@ fn main() -> Result<(), Box<dyn Error>> {
...
+            };
+        }
+
+        Ok(errors)
+    } ());
 
     // ...
 
-    Ok(())
+    std::process::exit(if !errors? { 0 } else { 1 });
```
Comment (tyranron) [inline 3_ecosystem/3_11_async/src/main.rs]: First, as we have rich type system, it's better to use appropriate type for that like [`std::process::ExitCode`](https://doc.rust-lang.org/stable/std/process/struct.ExitCode.html).

Second, as you're using [`std::process::Termination`](https://doc.rust-lang.org/stable/std/process/trait.Termination.html), then better stick with it to the end and don't mix exit code returning styles.

File: 3_ecosystem/3_11_async/src/main.rs [hunk 2]
```diff
@@ -41,46 +51,70 @@ fn main() -> Result<(), Box<dyn Error>> {
...
+                },
+                Some(Ok(url)) => {
+                    requests.push(request(&client, url));
+                },
+            };
 
-        for (url, filename) in urls {
-            tokio::spawn(request(&client, &url, filename));
+            if requests.len() >= 4 {
+                break;
```
Comment (tyranron) [inline 3_ecosystem/3_11_async/src/main.rs]: Why is that? 🤔

File: 3_ecosystem/3_11_async/src/main.rs [hunk 3]
```diff
@@ -41,46 +51,70 @@ fn main() -> Result<(), Box<dyn Error>> {
...
+            }
         }
 
-        // ...
+        loop {
+            match requests.next().await {
```
Comment (tyranron) [inline 3_ecosystem/3_11_async/src/main.rs]: Hmm... while it works, it's not much idiomatic as logic is not DRY and many concepts are re-implemented manually (like "at most 4 futures at the same time").

Consider to leverage existing abstractions. For example...

First, for convenience and combinators power we need to [convert `AsyncRead` into `Stream`](https://jsdw.me/posts/rust-futures-tokio/#converting-an-1) via something like that:
```bash 
let byte_stream = stream::unfold((), |_| {
    let next = io::read(io::stdin(), [0;1]).map(|(_stdin,buf,n)| (buf[0], ()));
    Some(next)
});
```

Second, on stream you can use [`.buffer_unordered()`](https://docs.rs/futures-preview/0.3.0-alpha.18/futures/stream/trait.StreamExt.html#method.buffer_unordered) to conveniently limit simultaneously processed futures.

And, finally, just [`.collect()`](https://docs.rs/futures-preview/0.3.0-alpha.18/futures/stream/trait.StreamExt.html#method.collect) your `Stream` to the `Vec` of `Result`s.

Comment (ffuugoo) [reply 3_ecosystem/3_11_async/src/main.rs]: > on stream you can use .buffer_unordered()

Reading documentation is hard. I've never noticed this one. 🤦

**Pull request merged**